lib/smart_answer/question/date.rb

```diff
@@ -50,7 +50,15 @@ def default_year(&block)
       end
 
       def range
-        @range ||= from && to ? from..to : false
+        if from && to
+          raise "to date must be after the from date" if from >= to
+
+          from..to
+        elsif !from && !to
+          false
+        else
+          raise "Both from and to must be defined to validate a date question"
+        end
```

Comment:
Nit: I think this will only be run at request time (when an input is being validated). It would be great if we could check the node definition at load time, so developers immediately know that definition is wrong. Could we put the checks in the initialiser?

Reply:
Is there an applicable place to do this prior to runtime without ending up in a 🐰 🕳️ ? I was under the impression that due to the DSL nature that you can essentially adjust the values at any point until runtime and by virtue of them being defined separately you can only handle all cases in this place (you could handle a little in the `to` method)

Reply:
🤦🏻 You're right, forgot that the `to`/`from` dates are defined as code blocks, which are resolved at request time, so won't be possible the validation at definition load time.

Just thinking how this error surfaces, I think it would be a 5xx page in production and the exception shown in development (?). Not much we can do about that I guess, except making sure we have good testing for the `to`/`from` blocks to help reduce chances of bugs.

Reply:
I wonder if we even need the range method? Could we simplify the validate_input method to:
```ruby
def validate_input(date)
  raise "to date must be after the from date" if from && to && from >= to

  if from && date < from || to && date > to
    raise InvalidResponse, "Provided date is out of range: #{date}", caller
  end
end
```

This would remove the constraint of requiring both `to` and `from` to be defined (not sure why we'd need it).

Reply:
Thanks, that's a great point that we no longer need a public range method.  You're right that it allows it to be simpler where we can either from, to or both and for the values to always have an effect. The concept of a range seems rooted in the past system of select options that needed a start and an end.

I wasn't too concerned about 5xx's in production as, given the values are rather static, I thought it was unlikely to pass through testing and more likely to just linger in the background as code that does nothing and remains. However with your suggested switch I think we can remove the exceptions altogether - if a developer sets up dates where `from > to` they will just create a node that cannot be progressed from - which should hopefully be caught in testing.

I'll just amend the commit now. Only lingering thought I have is that the code will need to be a bit more complex to show more relevant error messages - but that feels a challenge for another time (and also think it would warrant moving this code outside of the question - this seems to be the only question with built in validation - and into validate blocks)

Reply:
Agree with all those points! Thanks Kev, this looks great!